### ISSUE TITLE
Why is import hyperspy.api slow?

### ISSUE THREAD

**Issue by thomasaarholt** [hyperspy/hyperspy#1480]
Whilst I fully appreciate that part of the reason that hs is slow to import is that it hasn't been a priority to speed it up. It's fast enough. Still, considering that hyperspy only really needs a few dependencies (mentioned in #1438), and each of them imports almost instantaneously, I wonder why `hyperspy.api` takes longer. In fact, `import hyperspy.api` takes 3 times longer than `import numpy, scipy, matplotlib.pyplot, h5py, sklearn, IPython, traits, traitsui, ipyparallel`.

This isn't a big deal - I've just been restarting the kernel often and been wondering.
Perhaps some of the (optional) dependencies could be imported when necessary?

**Comment by to266** [comment]
To me it seems like (yet another) traits thing (I use TK because pyqt5 is broken, traits again).

(when just importing traits/traitsui I doubt those libraries do anything, whereas in our case they check/verify the local preferences that may also be read from the disk. `slugify` is also infamous for slowing things down, although at some point it went through a serious training regime to speed it up).


```python
In [1]: %matplotlib tk

In [2]: %prun import hyperspy.api as hs
         1548372 function calls (1499419 primitive calls) in 1.506 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     1542    0.171    0.000    0.171    0.000 {built-in method marshal.loads}
3432/3349    0.118    0.000    0.469    0.000 {built-in method builtins.__build_class__}
   2052/1    0.075    0.000    1.509    1.509 {built-in method builtins.exec}
   134/87    0.038    0.000    0.106    0.001 {built-in method _imp.create_dynamic}
    21906    0.034    0.000    0.068    0.000 utils.py:143(slugify)
59936/53901    0.032    0.000    0.041    0.000 {built-in method builtins.getattr}
     2462    0.032    0.000    0.032    0.000 {method '_trait' of 'traits.ctraits.CHasTraits' objects}
      269    0.024    0.000    0.088    0.000 has_traits.py:472(__init__)
     6855    0.022    0.000    0.022    0.000 {built-in method posix.stat}
    37193    0.021    0.000    0.031    0.000 {built-in method builtins.hasattr}
148842/148816    0.021    0.000    0.021    0.000 {built-in method builtins.isinstance}
     2677    0.020    0.000    0.082    0.000 <frozen importlib._bootstrap_external>:1233(find_spec)
      298    0.018    0.000    0.046    0.000 doccer.py:12(docformat)
      352    0.018    0.000    0.041    0.000 assumptions.py:313(__init__)
    14268    0.017    0.000    0.065    0.000 utils.py:334(__getattribute__)
     2225    0.017    0.000    0.126    0.000 <frozen importlib._bootstrap>:861(_find_spec)
 7638/276    0.017    0.000    0.115    0.000 utils.py:344(__setattr__)
     1542    0.016    0.000    0.024    0.000 <frozen importlib._bootstrap_external>:830(get_data)
     4248    0.014    0.000    0.014    0.000 {built-in method __new__ of type object at 0x7f8f4aefbce0}
  539/113    0.013    0.000    0.032    0.000 sre_parse.py:469(_parse)
    14638    0.010    0.000    0.027    0.000 <frozen importlib._bootstrap_external>:57(_path_join)
     1542    0.010    0.000    0.013    0.000 <frozen importlib._bootstrap_external>:430(_validate_bytecode_header)
     1542    0.009    0.000    0.241    0.000 <frozen importlib._bootstrap_external>:743(get_code)
     3084    0.009    0.000    0.022    0.000 <frozen importlib._bootstrap_external>:263(cache_from_source)
    14638    0.009    0.000    0.014    0.000 <frozen importlib._bootstrap_external>:59(<listcomp>)
     1208    0.009    0.000    0.021    0.000 inspect.py:2066(_signature_from_function)
   2328/1    0.009    0.000    1.509    1.509 <frozen importlib._bootstrap>:958(_find_and_load)
     1542    0.008    0.000    0.008    0.000 {method 'read' of '_io.FileIO' objects}
    21906    0.008    0.000    0.008    0.000 {method 'translate' of 'bytes' objects}
   2328/1    0.008    0.000    1.509    1.509 <frozen importlib._bootstrap>:931(_find_and_load_unlocked)
      633    0.008    0.000    0.096    0.000 has_traits.py:2350(on_trait_change)
     3608    0.008    0.000    0.011    0.000 <frozen importlib._bootstrap>:159(_get_module_lock)
    27310    0.008    0.000    0.008    0.000 {method 'replace' of 'str' objects}
   1680/2    0.008    0.000    1.508    0.754 <frozen importlib._bootstrap>:641(_load_unlocked)
14831/14731    0.007    0.000    0.009    0.000 logic.py:205(__hash__)
    49563    0.007    0.000    0.007    0.000 {method 'startswith' of 'str' objects}
     1097    0.007    0.000    0.024    0.000 trait_handlers.py:558(as_ctrait)
     3608    0.007    0.000    0.008    0.000 <frozen importlib._bootstrap>:77(acquire)
    13821    0.007    0.000    0.011    0.000 {built-in method builtins.setattr}
     1676    0.007    0.000    0.032    0.000 <frozen importlib._bootstrap>:493(_init_module_attrs)
     7319    0.007    0.000    0.007    0.000 {method 'splitlines' of 'str' objects}
      303    0.007    0.000    0.012    0.000 doccer.py:128(indentcount_lines)
    93075    0.007    0.000    0.007    0.000 {method 'append' of 'list' objects}
    26067    0.007    0.000    0.009    0.000 {method 'join' of 'str' objects}
     3608    0.006    0.000    0.007    0.000 <frozen importlib._bootstrap>:102(release)
    11054    0.006    0.000    0.015    0.000 has_traits.py:169(_get_def)
      762    0.006    0.000    0.006    0.000 {method 'sub' of '_sre.SRE_Pattern' objects}
 1056/762    0.006    0.000    0.057    0.000 traits_listener.py:339(register)
      333    0.006    0.000    0.008    0.000 sre_compile.py:250(_optimize_charset)
1848/1328    0.006    0.000    0.043    0.000 inspect.py:2147(_signature_from_callable)
        3    0.006    0.002    0.006    0.002 {built-in method posix.read}
     2225    0.006    0.000    0.097    0.000 <frozen importlib._bootstrap_external>:1117(_get_spec)
    14740    0.006    0.000    0.006    0.000 sre_parse.py:232(__next)
    22664    0.006    0.000    0.006    0.000 {method 'encode' of 'str' objects}
      544    0.006    0.000    0.006    0.000 {method 'findall' of '_sre.SRE_Pattern' objects}
    34570    0.006    0.000    0.006    0.000 {method 'rstrip' of 'str' objects}
 11323/10    0.006    0.000    1.508    0.151 <frozen importlib._bootstrap>:989(_handle_fromlist)
        3    0.005    0.002    0.005    0.002 {built-in method _posixsubprocess.fork_exec}
     1531    0.005    0.000    0.007    0.000 inspect.py:2679(__init__)
    34868    0.005    0.000    0.005    0.000 {method 'get' of 'dict' objects}
    21906    0.005    0.000    0.005    0.000 {built-in method unicodedata.normalize}
54747/54122    0.005    0.000    0.011    0.000 {built-in method builtins.len}
    18648    0.005    0.000    0.005    0.000 {method 'rpartition' of 'str' objects}
   1905/2    0.005    0.000    0.115    0.058 utils.py:231(add_dictionary)
  963/103    0.005    0.000    0.024    0.000 sre_compile.py:64(_compile)
     2898    0.005    0.000    0.005    0.000 {method 'format' of 'str' objects}
     1676    0.005    0.000    0.008    0.000 <frozen importlib._bootstrap>:304(__exit__)
```

**Comment by sem-geologist** [comment]
I guess because there is circular imports, landscape complains about it.

**Comment by francisco-dlp** [comment]
Good point, would you like to open an issue about it?

**Comment by thomasaarholt** [comment]
Doing a bit of investigation:
Importing the api takes 1970ms 
Of this, the api_no_gui takes by far the longest, at 1880ms.

Within this, BaseSignal takes 1770ms, Signal1D takes 238ms and EDSSpectrum takes 208ms. EELSSpectrum takes 15ms.

**Comment by francisco-dlp** [comment]
And what takes so long in BaseSignal?

**Comment by thomasaarholt** [comment]
~~It's all with inheriting the properties of MVA, MVATools and FancySlicing.~~

Actually, no. I was making an assumption there.

It's to do with the imports. mpl isn't too bad, at 200ms. The io and drawing packages however, are at 500ms and 440ms respectively. I'm gonna take a look and see if any of the loading packages use any very general import statements, like import hyperspy.api as hs.

Amongst the imports of hyperspy.signal:
```
numpy&scipy 0.06500005722045898
mpl 0.19299983978271484
traits&numbers 0.043000221252441406
axes 0.059999704360961914
io 0.49500012397766113
drawing 0.44300007820129395
mva 0.28999996185302734
```


Edit2:
Looking into io: The io.py import statements take up 99% of the time.

**The tiff and EMD are particularly slow at 144ms and 89ms respectively.**  Looks like the tiff one is almost entirely from TiffFile though.

**Comment by thomasaarholt** [comment]
Question: Is it necessary/desireable to load the file readers before we need them?

**Comment by thomasaarholt** [comment]
I've just spent some time looking at where `import hyperspy.api` is slow. I do this using `tuna` and the `-X importtime` python argument, to generate output that is more easily parsable to my human eyes. On my reasonably beefy windows computer, it takes 5 seconds to import hyperspy.api.

```bash
python -X importtime -c "import hyperspy.api" 2> hs.txt && tuna hs.txt
```
<table>
<tbody>
<tr>
<td>
<p float="left">
<img src="https://user-images.githubusercontent.com/2721423/85368062-cf561300-b52a-11ea-8e4b-ff717b1b2287.png" width="600" />
</td>

<td>

<table>
<thead>
<tr>
<th>package</th>
<th>time ms</th>
</tr>
</thead>
<tbody>
<tr><td>traitsui</td> <td>950</td></tr>
<tr><td>sympy</td> <td>482</td></tr>
<tr><td>numba</td> <td>458</td></tr>
<tr><td>sklearn</td> <td>294</td></tr>
<tr><td>pint</td> <td>293</td></tr>
<tr><td>dask</td> <td>239</td></tr>
<tr><td>matplotlib</td> <td>215</td></tr>
<tr><td>scipy.signal</td><td>208</td> </tr>
<tr><td>numpy</td> <td>171</td></tr>
<tr><td>ipywidgets</td> <td>113</td></tr>
<tr><td>h5py</td> <td>109</td></tr>
</td>
</tr>
</tbody>
</table>
</table>

It is clear that if we can avoid importing traitsui, sympy and numba at hyperspy startup, we can make hyperspy about 2 seconds faster to load.

**Comment by francisco-dlp** [comment]
That's really useful. The best way not to import traitsui is disabling it in preferences or simply not installing the hyperspy traitsui widgets. The fact that using traitsui slows add a 1s delay should be documented somewhere.

 As for ``sympy`` and ``numba``, we could lazy-import them.

I think that we should have a FAQs section somewhere. Maybe just a FAQs label in GitHub would do or somebody has a better idea?

**Comment by tjof2** [comment]
FWIW, I managed to get the `hyperspy.api_nogui` import time down from 2.21s on my machine to 1.45s, by moving a number of import statements inside relevant functions. But it was actually quite a lot of work (more than one would expect) and touched a lot of files (30+).

The biggest external culprits are now:

- `pint` (0.16s)
- `matplotlib` (0.15s)
- `sklearn` (0.15s)
- `dask` (0.12s)
- `pkg_resources` (0.12s) <- only used in `hyperspy/extensions.py`. It's a common culprit for slow imports

If you have `pyUSID` installed too, then that is particularly bad - it takes 0.374s to import (or 20% of the total).

**Comment by thomasaarholt** [comment]
Hehe, I did exactly the same! Pint was a bit of a pain - I tried moving inside DataAxis (init), but DataAxis apparently gets called a __lot__, and made the tests that I ran just before bed last night take over an hour! So I couldn't hide that import.

I managed to hide sklearn too. 

@francisco-dlp's suggestion of lazy imports is probably the best one for long term, but hiding the imports inside functions would probably work just as well for most functions. - We already do this quite a bit here and there.

We should only import traitsui if it is enabled in the prefs. Currently it imports it anyway.

**Comment by tjof2** [comment]
That's interesting - I didn't see any slowdown in the full test suite (took ~400s both times), though I didn't touch `pint`. Also these SO questions are relevant: 

- https://stackoverflow.com/questions/32151193/is-there-a-performance-cost-putting-python-imports-inside-functions
- https://stackoverflow.com/questions/296036/does-python-optimize-modules-when-they-are-imported-multiple-times

But putting them inside functions etc was messy, only saved ~30%, and isn't very friendly for future development and maintenance. Agree that lazy-imports (particularly things like sklearn) are the way to go.

**Comment by thomasaarholt** [comment]
The slowdown was only due to placing `import pint` and a `self._ureg = pint.UnitRegistry()` under `DataAxis.__init__`.  It turns out that this class is called **a lot**, and the latter call takes about 130ms.

**Comment by thomasaarholt** [comment]
Having another look at this now, using https://github.com/xonsh/lazyasd for the lazy imports (it's on conda and pypi).

A good trick to ensure that one is actually loading modules lazily is to temporarily stick a `raise ValueError('lazy')` in the top `__init__.py` files of the modules. That way I noticed that dask is implicitly importing numba - and we import `a lot` of different things from dask, which makes it a bit tricky to lazily load, for instance. So for the moment, numba and dask are out.
(but I've shaved ~1sec off import times on my windows just by lazy loading pint).
